File: JAVA/lld-examples/hard/distributed-chat-system/src/main/java/com/systemdesign/lld/chat/model/ChatRoom.java
package com.systemdesign.lld.chat.model;

import java.time.LocalDateTime;
import java.util.*;

/**
 * ChatRoom model representing a chat room in the distributed system
 */
public class ChatRoom {
    private final String roomId;
    private String name;
    private String description;
    private final String creatorId;
    private final LocalDateTime createdAt;
    private boolean isActive;
    private boolean isPrivate;
    private int maxMembers;
    
    // Members and roles
    private Set<String> members; // Set of user IDs
    private Set<String> admins; // Set of admin user IDs
    private Set<String> bannedUsers; // Set of banned user IDs
    
    // Room settings
    private boolean isReadOnly;
    private int messageRetentionDays;
    private boolean allowFileSharing;
    private boolean allowReactions;
    
    // Statistics
    private int totalMessages;
    private LocalDateTime lastActivity;
    
    public ChatRoom(String name, String description, User creator) {
        this.roomId = UUID.randomUUID().toString();
        this.name = name;
        this.description = description;
        this.creatorId = creator.getUserId();
        this.createdAt = LocalDateTime.now();
        this.isActive = true;
        this.isPrivate = false;
        this.maxMembers = 1000;
        
        // Initialize collections
        this.members = new HashSet<>();
        this.admins = new HashSet<>();
        this.bannedUsers = new HashSet<>();
        
        // Default settings
        this.isReadOnly = false;
        this.messageRetentionDays = 30;
        this.allowFileSharing = true;
        this.allowReactions = true;
        
        // Statistics
        this.totalMessages = 0;
        this.lastActivity = LocalDateTime.now();
        
        // Add creator as admin and member
        this.admins.add(creator.getUserId());
        this.members.add(creator.getUserId());
    }
    
    /**
     * Add member to room
     */
    public boolean addMember(String userId) {
        if (bannedUsers.contains(userId)) {
            return false;
        }
        
        if (members.size() >= maxMembers) {
            return false;
        }
        
        members.add(userId);
        lastActivity = LocalDateTime.now();
        return true;
    }
    
    /**
     * Remove member from room
     */
    public boolean removeMember(String userId) {
        if (members.contains(userId)) {
            members.remove(userId);
            // Remove from admins if they were admin
            admins.remove(userId);
            lastActivity = LocalDateTime.now();
            return true;
        }
        return false;
    }
    
    /**
     * Add admin to room
     */
    public boolean addAdmin(String userId) {
        if (members.contains(userId)) {
            admins.add(userId);
            return true;
        }
        return false;
    }
    
    /**
     * Remove admin from room
     */
    public boolean removeAdmin(String userId) {
        if (admins.contains(userId) && !userId.equals(creatorId)) {
            admins.remove(userId);
            return true;
        }
        return false;
    }
    
    /**
     * Ban user from room
     */
    public boolean banUser(String userId) {
        if (userId.equals(creatorId)) {
            return false;
        }
        
        bannedUsers.add(userId);
        removeMember(userId);
        return true;
    }
    
    /**
     * Unban user from room
     */
    public boolean unbanUser(String userId) {
        return bannedUsers.remove(userId);
    }
    
    /**
     * Check if user is a member
     */
    public boolean isMember(String userId) {
        return members.contains(userId);
    }
    
    /**
     * Check if user is an admin
     */
    public boolean isAdmin(String userId) {
        return admins.contains(userId);
    }
    
    /**
     * Check if user is the creator
     */
    public boolean isCreator(String userId) {
        return userId.equals(creatorId);
    }
    
    /**
     * Check if user is banned
     */
    public boolean isBanned(String userId) {
        return bannedUsers.contains(userId);
    }
    
    /**
     * Check if user can send messages
     */
    public boolean canSendMessage(String userId) {
        return isMember(userId) && 
               !isBanned(userId) && 
               (!isReadOnly || isAdmin(userId));
    }
    
    /**
     * Get member count
     */
    public int getMemberCount() {
        return members.size();
    }
    
    /**
     * Get admin count
     */
    public int getAdminCount() {
        return admins.size();
    }
    
    /**
     * Update activity
     */
    public void updateActivity() {
        this.lastActivity = LocalDateTime.now();
    }
    
    /**
     * Increment message count
     */
    public void incrementMessageCount() {
        this.totalMessages++;
        updateActivity();
    }
    
    /**
     * Set room as read-only
     */
    public void setReadOnly(boolean readOnly) {
        this.isReadOnly = readOnly;
    }
    
    /**
     * Set room as private
     */
    public void setPrivate(boolean isPrivate) {
        this.isPrivate = isPrivate;
    }
    
    /**
     * Update room description
     */
    public void updateDescription(String description) {
        this.description = description;
    }
    
    /**
     * Deactivate room
     */
    public void deactivate() {
        this.isActive = false;
    }
    
    /**
     * Activate room
     */
    public void activate() {
        this.isActive = true;
    }
    
    /**
     * Get room information
     */
    public Map<String, Object> getRoomInfo() {
        Map<String, Object> info = new HashMap<>();
        info.put("roomId", roomId);
        info.put("name", name);
        info.put("description", description);
        info.put("creatorId", creatorId);
        info.put("createdAt", createdAt);
        info.put("isActive", isActive);
        info.put("isPrivate", isPrivate);
        info.put("isReadOnly", isReadOnly);
        info.put("memberCount", getMemberCount());
        info.put("adminCount", getAdminCount());
        info.put("totalMessages", totalMessages);
        info.put("lastActivity", lastActivity);
        info.put("maxMembers", maxMembers);
        info.put("messageRetentionDays", messageRetentionDays);
        info.put("allowFileSharing", allowFileSharing);
        info.put("allowReactions", allowReactions);
        return info;
    }
    
    // Getters
    public String getRoomId() { return roomId; }
    public String getName() { return name; }
    public String getDescription() { return description; }
    public String getCreatorId() { return creatorId; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public boolean isActive() { return isActive; }
    public boolean isPrivate() { return isPrivate; }
    public int getMaxMembers() { return maxMembers; }
    public Set<String> getMembers() { return new HashSet<>(members); }
    public Set<String> getAdmins() { return new HashSet<>(admins); }
    public Set<String> getBannedUsers() { return new HashSet<>(bannedUsers); }
    public boolean isReadOnly() { return isReadOnly; }
    public int getMessageRetentionDays() { return messageRetentionDays; }
    public boolean isAllowFileSharing() { return allowFileSharing; }
    public boolean isAllowReactions() { return allowReactions; }
    public int getTotalMessages() { return totalMessages; }
    public LocalDateTime getLastActivity() { return lastActivity; }
    
    // Setters
    public void setName(String name) { this.name = name; }
    public void setMaxMembers(int maxMembers) { this.maxMembers = maxMembers; }
    public void setMessageRetentionDays(int messageRetentionDays) { this.messageRetentionDays = messageRetentionDays; }
    public void setAllowFileSharing(boolean allowFileSharing) { this.allowFileSharing = allowFileSharing; }
    public void setAllowReactions(boolean allowReactions) { this.allowReactions = allowReactions; }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ChatRoom chatRoom = (ChatRoom) o;
        return Objects.equals(roomId, chatRoom.roomId);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(roomId);
    }
    
    @Override
    public String toString() {
        return "ChatRoom{" +
                "roomId='" + roomId + '\'' +
                ", name='" + name + '\'' +
                ", members=" + members.size() +
                ", active=" + isActive +
                '}';
    }
} 

File: Python/lld-examples/hard/distributed-chat-system/chat_room.py
"""
ChatRoom class for the distributed chat system
"""

import uuid
from datetime import datetime
from typing import Set, Optional
from user import User


class ChatRoom:
    """ChatRoom class representing a chat room"""
    
    def __init__(self, name: str, description: str, creator: User):
        self.room_id = str(uuid.uuid4())
        self.name = name
        self.description = description
        self.creator_id = creator.user_id
        self.created_at = datetime.now()
        self.is_active = True
        self.is_private = False
        self.max_members = 1000
        
        # Members and roles
        self.members: Set[str] = set()  # Set of user IDs
        self.admins: Set[str] = set()  # Set of admin user IDs
        self.banned_users: Set[str] = set()  # Set of banned user IDs
        
        # Room settings
        self.is_read_only = False
        self.message_retention_days = 30
        self.allow_file_sharing = True
        self.allow_reactions = True
        
        # Statistics
        self.total_messages = 0
        self.last_activity = datetime.now()
        
        # Add creator as admin
        self.admins.add(creator.user_id)
        self.members.add(creator.user_id)
    
    def add_member(self, user_id: str) -> bool:
        """Add member to room"""
        if user_id in self.banned_users:
            return False
        
        if len(self.members) >= self.max_members:
            return False
        
        self.members.add(user_id)
        self.last_activity = datetime.now()
        return True
    
    def remove_member(self, user_id: str) -> bool:
        """Remove member from room"""
        if user_id in self.members:
            self.members.remove(user_id)
            # Remove from admins if they were admin
            if user_id in self.admins:
                self.admins.remove(user_id)
            self.last_activity = datetime.now()
            return True
        return False
    
    def add_admin(self, user_id: str) -> bool:
        """Add admin to room"""
        if user_id in self.members:
            self.admins.add(user_id)
            return True
        return False
    
    def remove_admin(self, user_id: str) -> bool:
        """Remove admin from room"""
        if user_id in self.admins and user_id != self.creator_id:
            self.admins.remove(user_id)
            return True
        return False
    
    def ban_user(self, user_id: str) -> bool:
        """Ban user from room"""
        if user_id == self.creator_id:
            return False
        
        self.banned_users.add(user_id)
        self.remove_member(user_id)
        return True
    
    def unban_user(self, user_id: str) -> bool:
        """Unban user from room"""
        if user_id in self.banned_users:
            self.banned_users.remove(user_id)
            return True
        return False
    
    def is_member(self, user_id: str) -> bool:
        """Check if user is a member"""
        return user_id in self.members
    
    def is_admin(self, user_id: str) -> bool:
        """Check if user is an admin"""
        return user_id in self.admins
    
    def is_creator(self, user_id: str) -> bool:
        """Check if user is the creator"""
        return user_id == self.creator_id
    
    def is_banned(self, user_id: str) -> bool:
        """Check if user is banned"""
        return user_id in self.banned_users
    
    def can_send_message(self, user_id: str) -> bool:
        """Check if user can send messages"""
        return (self.is_member(user_id) and 
                not self.is_banned(user_id) and 
                (not self.is_read_only or self.is_admin(user_id)))
    
    def get_member_count(self) -> int:
        """Get member count"""
        return len(self.members)
    
    def get_admin_count(self) -> int:
        """Get admin count"""
        return len(self.admins)
    
    def update_activity(self):
        """Update last activity"""
        self.last_activity = datetime.now()
    
    def increment_message_count(self):
        """Increment message count"""
        self.total_messages += 1
        self.update_activity()
    
    def set_read_only(self, read_only: bool):
        """Set room as read-only"""
        self.is_read_only = read_only
    
    def set_private(self, private: bool):
        """Set room as private"""
        self.is_private = private
    
    def update_description(self, description: str):
        """Update room description"""
        self.description = description
    
    def deactivate(self):
        """Deactivate room"""
        self.is_active = False
    
    def activate(self):
        """Activate room"""
        self.is_active = True
    
    def get_room_info(self) -> dict:
        """Get room information"""
        return {
            'room_id': self.room_id,
            'name': self.name,
            'description': self.description,
            'creator_id': self.creator_id,
            'created_at': self.created_at.isoformat(),
            'is_active': self.is_active,
            'is_private': self.is_private,
            'is_read_only': self.is_read_only,
            'member_count': self.get_member_count(),
            'admin_count': self.get_admin_count(),
            'total_messages': self.total_messages,
            'last_activity': self.last_activity.isoformat(),
            'max_members': self.max_members,
            'message_retention_days': self.message_retention_days,
            'allow_file_sharing': self.allow_file_sharing,
            'allow_reactions': self.allow_reactions
        }
    
    def __str__(self):
        return f"ChatRoom(name={self.name}, members={len(self.members)}, active={self.is_active})"
    
    def __repr__(self):
        return self.__str__() 